### PR TITLE
Avoid selection of 0 propensity reactions and negative counts

### LINKED ISSUE #48: Arrow hangs wcEcoli complexation with certain sim seeds
Some wcEcoli sims can hang during complexation.

See https://github.com/CovertLab/wcEcoli/issues/1229 including @tahorst's boiled down [test case](https://github.com/CovertLab/wcEcoli/blob/arrow-hang/arrow_hang.py).

### PR DESCRIPTION
**Pull request by tahorst**
This fixes #48 by avoiding reactions with a propensity of 0 when making the random choice.  The issue came from the random number being exactly 0 which meant the first reaction would always be selected even if the propensity is 0 since `point` was 0.  This caused negative counts since the reaction did not have counts to proceed which then gave a positive propensity for negative counts since they formed a dimer.

I've also added a check to cause a failure if we hit negative counts instead of an infinite memory sucking loop and rearranged some logic control flow that didn't quite make sense (checking `choice == -1` after `choice` is set to 0 and incremented and now adding a break on an overflow error).

@1fish2: not sure if you wanted the `arrow-hang` branch committed to master yet but feel free to add here (or merge afterwards) since the test now passes.

### CODE REVIEW

**Review by 1fish2 — APPROVED** [review]
This is great!

If you're ready, go ahead and merge or squash-merge.

@U8NWXD, @prismofeverything `test_flagella` fails and was failing in the previous release. It raises `SimulationFailure`. Any guidance? Can anyone fix it quickly before doing a release for Travis to use in ecoli?

@U8NWXD Do you have a release checklist? I'm thinking:
* make; pytest; bump the version number [done]
* make a release tag
* build a pip and push it to PyPI [I forget how]

**Comment by U8NWXD** [comment]
@1fish2 I don't have any release procedures for arrow, but in Vivarium Core I wrote a [simple bash script](https://github.com/vivarium-collective/vivarium-core/blob/master/release.sh) to automated most of the release process. It:

* Requires that the user specify what version they want to create. This forces the user to double-check that they're releasing the correct version. An improvement might be to check this number, e.g. to ensure that version numbers are semver-compliant.
* Checks that the `setup.py` version matches the version passed to the release script. An improvement might be to also check that a changelog has been updated.
* Checks that the user is currently on `master` and has a clean working directory. This ensures that the code release to PyPI is the same as the code we assign a git tag (below).
* Creates and pushes a git tag with the version number.
* Creates and uploads a source code release to PyPI.

It doesn't check that the tests are passing because in Vivarium Core, we expect the tests to always pass on `master`. I'd recommend setting up GitHub actions to make sure tests always pass on `master` here, but you could also put running the tests into the release script as an alternative.

I don't have any insights on `test_flagella`. @prismofeverything it looks like you introduced that test. If you have a chance to look at it, do you have any ideas about what's going wrong?

**Comment by 1fish2** [comment]
@U8NWXD Cool. I'll crib from that script. Borealis has a simpler script + manual steps including uploading to PyPI in case one wants to decide if it's ready to release.

**Comment by 1fish2** [comment]
@U8NWXD `test_memory` fails when run by itself and sometimes when run within `test_arrow`. It's trying to catch native malloc leaks with `psutil.Process(os.getpid()).memory_info().rss`.

Did you find a workable technique in the swiglpk memory experiments? Does tracemalloc trace allocations in C code?

**Comment by U8NWXD** [comment]
@1fish2 For my swiglpk debugging, I used `memory_profiler`. It seemed to record allocations from swiglpk, but it also didn't show all the deallocations I expected. I never managed to untangle whether that was just how Python's garbage collector worked or whether there was a C-level memory leak

File: arrow/obsidian.c [hunk 1]
```diff
@@ -208,21 +209,21 @@ evolve_result evolve(Info *info, double duration, int64_t *state, double *rates)
...
+        break;
+      }
+
       // Based on the random sample, find the event that it corresponds to by
       // iterating through the propensities until we surpass our sampled value
       choice = 0;
       progress = 0.0;
 
-      while (progress + propensities[choice] < point) {
+      while (progress + propensities[choice] < point || propensities[choice] == 0) {
```
[the remark this answers is not in the document]
Comment (tahorst) [reply arrow/obsidian.c:236]: Good thinking!  I like the safety checks.

**Pull request merged**